WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/MainActivity.kt

```diff
@@ -0,0 +1,332 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.android.wearable.speaker
+
+import android.Manifest
+import android.content.pm.PackageManager
+import android.media.AudioDeviceInfo
+import android.media.AudioManager
+import android.media.MediaPlayer
+import android.os.Bundle
+import android.util.Log
+import android.widget.Toast
+import androidx.activity.result.contract.ActivityResultContracts.RequestPermission
+import androidx.appcompat.content.res.AppCompatResources
+import androidx.constraintlayout.motion.widget.MotionLayout
+import androidx.core.content.ContextCompat
+import androidx.core.content.getSystemService
+import androidx.fragment.app.FragmentActivity
+import androidx.lifecycle.lifecycleScope
+import androidx.wear.ambient.AmbientModeSupport
+import com.example.android.wearable.speaker.databinding.MainActivityBinding
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import java.time.Duration
+import kotlin.coroutines.resume
+
+/**
+ * We first get the required permission to use the MIC. If it is granted, then we continue with
+ * the application and present the UI with three icons: a MIC icon (if pressed, user can record up
+ * to 10 seconds), a Play icon (if clicked, it wil playback the recorded audio file) and a music
+ * note icon (if clicked, it plays an MP3 file that is included in the app).
+ */
+class MainActivity : FragmentActivity(), AmbientModeSupport.AmbientCallbackProvider {
+
+    /**
+     * Ambient mode controller attached to this display. Used by Activity to see if it is in
+     * ambient mode.
+     */
+    private lateinit var ambientController: AmbientModeSupport.AmbientController
+
+    private lateinit var binding: MainActivityBinding
+
+    private var soundRecorder: SoundRecorder? = null
+
+    private var appState: AppState = AppState.READY
+
+    private var ongoingWork: Job = Job().apply { complete() }
+
+    private val requestPermissionLauncher = registerForActivityResult(RequestPermission()) { isGranted ->
+        if (isGranted) {
+            start()
+        } else {
+            // Permission has been denied before. At this point we should show a dialog to
+            // user and explain why this permission is needed and direct them to go to the
+            // Permissions settings for the app in the System settings. For this sample, we
+            // simply exit to get to the important part.
+            Toast.makeText(this, R.string.exiting_for_permissions, Toast.LENGTH_LONG).show()
+            finish()
```

Comment:
We shouldn't finish the app, because you don't generally want to close app if a permission is denied. Also, they can still listen to music if they want without mic permissions.

Reply:
Updated in 791ae1b!